grr/server/grr_response_server/gui/ui/components/flow_details/helpers/osquery_test_utils.ts

```diff
@@ -0,0 +1,208 @@
+import { OsqueryColumn, OsqueryRow } from '@app/lib/api/api_interfaces';
+import { newFlowListEntry, newFlowResultSet } from '@app/lib/models/model_test_util';
+import { DeepMutable } from '@app/lib/type_utils';
+import { FlowListEntry, FlowState, FlowResultSet } from '@app/lib/models/flow';
+import { DebugElement } from '@angular/core';
+import { By } from '@angular/platform-browser';
+
+type FullMutableOsqueryTable = {
+  query: string,
+  header: {
+    columns: ReadonlyArray<OsqueryColumn>,
+  },
+  rows: ReadonlyArray<OsqueryRow>
+};
+
+function newOsqueryTable(): FullMutableOsqueryTable {
+  return {
+    query: '',
+    header: {
+      columns: [] as ReadonlyArray<OsqueryColumn>,
+    },
+    rows: [] as ReadonlyArray<OsqueryRow>,
+  };
+}
+
+export class OsqueryTableBuilder {
+  private table?: FullMutableOsqueryTable;
+
+  withColumns(columns: ReadonlyArray<string>): OsqueryTableBuilder {
+    if (this.table === undefined) {
+      this.table = newOsqueryTable();
+    }
+
+    this.table.header.columns = columns.map((colName) => ({name: colName}));
+    return this;
+  }
+
+  withValues(values: ReadonlyArray<ReadonlyArray<string>>): OsqueryTableBuilder {
+    if (this.table === undefined) {
+      this.table = newOsqueryTable();
+    }
+
+    this.table.rows = values.map((rowValues) => ({values: rowValues}));
+    return this;
+  }
+
+  withQuery(query: string) : OsqueryTableBuilder {
+    if (this.table === undefined) {
+      this.table = newOsqueryTable();
+    }
+
+    this.table.query = query;
+
+    return this;
+  }
+
+  withQueryIfDefined(query: string): OsqueryTableBuilder {
+    if (this.table !== undefined) {
+      this.table.query = query;
+    }
+
+    return this;
+  }
+
+  build(): FullMutableOsqueryTable | undefined {
+    return this.table;
+  }
+
+  buildWithQueryIfDefined(query: string): FullMutableOsqueryTable | undefined {
+    this.withQueryIfDefined(query);
+    return this.build();
+  }
+}
+
+/** Helper class to build a FlowListEntry objects in a declarative manner */
+export class OsqueryFlowListEntryBuilder {
```

Comment:
The following code

```
const testFlowListEntry = new OsqueryFlowListEntryBuilder()
  .withFlowState(FlowState.FINISHED)
  .withProgressTable([testColName], [[testCellValue]])
  .withProgressRowsCount(1)
  .withQuery(testQuery);
```

should effectively be equivalent to:
```
{
  ...newFlowListEntry({
    state: FlowState.FINISHED,
    progress: {
      totalRowsCount: 1,
      partialTable: {
        ...EMPTY_OSQUERY_TABLE,
        header: {
          [testColName],
        },
        rows: [{values: [testCellValue]}]
      }
    }
  }),
  
  query: testQuery,
}
```

This is slightly more complicated than the builder variant but also much more explicit and you don't need the 150 lines of untested (and fairly complicated) builder code. 

Probably we can define newOsqueryResultsTable(t: Partial<OsqueryResultsTable>) helper function to save a few lines of code, but I wouldn't go far beyond that. TypeScript's ellipsis (`...`) operator - something that doesn't exist in Java (last time I checked) should make building new test objects generally easy and reduces the need for the super-verbose builder pattern.

Reply:
Yeap I see your point; I will refactor this and remove the builder. But can we appreciate for the last time how beautiful, declarative and concise tests are? I'm probably too attached to this code.